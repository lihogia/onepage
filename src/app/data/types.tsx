
export interface UtilLink {
    title: string,
    url: string
}

export interface SimpleSearch extends UtilLink {
    fieldname: string,
    pattern: string
}

export type Util = UtilLink | SimpleSearch;

export function getUtilTypeName(util: Util) {
    if ('fieldname' in util) {
        return 'SimpleSearch';
    }else {
        return 'UtilLink';
    }
}


export interface SubCategory {
    name: string,
    utils: Util[]
}

export interface Category {
    name: string,
    subcategories: SubCategory[]
}

export interface BoardSettings {
    categories: Category[],
    selectedIndex: number,
    mode: number,
    contextMenus: Map<string, boolean>, // true: show, false: hide the context Menu, key is the id of the Context Menu Item List
}

export interface OnePageSettings {
    categories: Category[],
    version: string
}