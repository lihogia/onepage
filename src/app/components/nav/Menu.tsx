
import Image from 'next/image';
import { useContext, useState } from 'react';
import type { Category } from '@/app/data/types';
import { BoardContext } from '@/app/components/BoardContext';
import CategoryOnMenu from '@/app/components/categoryOnMenu';

export default function Menu(
    {categories, selectedIndex}: 
    {categories: Category[], selectedIndex: number}) {

    const boardContext = useContext(BoardContext);
    const isEdit = boardContext.isEdit();
    const isConfig = boardContext.boardSettings.mode === 3;
    const isAbout = boardContext.boardSettings.mode === 2;
    const isDonate = boardContext.boardSettings.mode === 4;

    /** Use in mobile view */
    function openMenu(isOpen: boolean) {
        const containerElement = document.getElementById('ContainerID');
        if (containerElement != null) {
            containerElement.className = isOpen ? 'container_6' : 'container';
        }

        const menuElement = document.getElementById('MenuID');
        if (menuElement != null) {
            menuElement.className = isOpen ? 'grid1m_sub' : 'grid1m_sub_none';
        }
    }

    const switchMode = (mode: number) => {
        boardContext.setMode(mode);
    }

    return (
    <>
    <div className={'grid1'}>
        <section className="logo"><Image src='/onepage.png' alt='OnePage' width="90" height="90" priority={true}/></section>
        <ul className="menu">
            {categories.map((element, index) => {
                 if (index === selectedIndex) {
                    return (
                        <CategoryOnMenu category={element} index={index} key={`${index}_${element.name}`} isMobile={false} isLast={categories.length === 1}/>
                    );
                }else {
                    return (
                        <li key={`${index}_${element.name}`} className='menuItem'><a href="#" onClick={
                            () => {
                                boardContext.setSelectedCategoryIndex(index);
                            }
                        }>{element.name}</a></li>    
                    );    
                }
            })
            }
        </ul>
        <ul className="menuBottom">
            {!isAbout && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(2) }}>About</a></li>}
            {isAbout && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to Categories</a></li>}

            {!isEdit && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(1) }}>Edit</a></li>}
            {isEdit && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to View</a></li>}

            {!isConfig && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(3) }}>Config</a></li>}
            {isConfig && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to Categories</a></li>}

            <li className="menuItemBottom"><a href="https://www.buymeacoffee.com/lilogia" target="_blank">Buy me a coffee</a></li>
            {/*!isDonate && <li className="menuItemBottom"><a href="#" onClick={() => { boardContext.setMode(4) }}>Donate</a></li>*/}
            {/*isDonate && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { boardContext.setMode(0) }}>Back to Categories</a></li>*/}
        </ul>    
    </div>
    <div className="grid1m">
        <section className="logo"><Image src='/onepage.png' alt='OnePage' width="90" height="90" priority={true}/></section>
        <ul className="menuTitle">
            <li><a href="#" onClick={() => {
                openMenu(true);
            }}>Categories</a></li>
        </ul>
    </div>
    <div className="grid1m_sub_none" id="MenuID">
        <ul className="menu">
            {categories.map((element, index) => {
                 if (index === selectedIndex) {
                    return (
                        <CategoryOnMenu category={element} index={index} key={`${index}_${element.name}`} isMobile={true} isLast={categories.length === 1}/>
                    );
                }else {
                    return (
                        <li key={`${index}_${element.name}`} className='menuItem'><a href="#" onClick={
                            () => {
                                if (!isEdit) {
                                    openMenu(false);
                                }
                                boardContext.setSelectedCategoryIndex(index);
                            }
                        }>{element.name}</a></li>    
                    );    
                }
                
            })
            }

        </ul>
        <ul className="menuBottom">
            {!isAbout && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(2) }}>About</a></li>}
            {isAbout && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to Categories</a></li>}

            {!isEdit && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(1) }}>Edit</a></li>}
            {isEdit && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to View</a></li>}

            {!isConfig && <li className="menuItemBottom"><a href="#" onClick={() => { switchMode(3) }}>Config</a></li>}
            {isConfig && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { switchMode(0) }}>Back to Categories</a></li>}

            <li className="menuItemBottom"><a href="https://www.buymeacoffee.com/lilogia" target="_blank">Buy me a coffee</a></li>
            {/*!isDonate && <li className="menuItemBottom"><a href="#" onClick={() => { boardContext.setMode(4) }}>Donate</a></li>*/}
            {/*isDonate && <li className="menuItemBottom"><a href="#" className='menuItemBottomSelected' onClick={() => { boardContext.setMode(0) }}>Back to Categories</a></li>*/}

        </ul>
    </div>
    </>
    );

}

/*

*/