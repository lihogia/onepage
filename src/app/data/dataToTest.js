export const test001 = {
    "categories": [
        {
            "name": "Common",
            "subcategories": [
                {
                    "name": "Office Tools",
                    "utils": [
                        {
                            "title": "Google Mail",
                            "url": "https://mail.google.com/mail/u/0/"
                        },
                        {
                            "title": "Microsoft Outlook",
                            "url": "https://outlook.live.com/mail/0/"
                        },
                        {
                            "title": "Yahoo! Mail",
                            "url": "https://mail.yahoo.com/"
                        }
                    ]
                },
                {
                    "name": "Search",
                    "utils": [
                        {
                            "fieldname": "q",
                            "title": "Google",
                            "url": "https://www.google.com/search"
                        },
                        {
                            "fieldname": "q",
                            "title": "Microsoft",
                            "url": "https://www.bing.com/search"
                        },
                        {
                            "title": "百度",
                            "url": "https://www.baidu.com/"
                        }
                    ]
                }
            ]
        },
        {
            "name": "News",
            "subcategories": [
                {
                    "name": "US",
                    "utils": [
                        {
                            "title": "CNN",
                            "url": "https://edition.cnn.com/"
                        },
                        {
                            "title": "npr",
                            "url": "https://www.npr.org/"
                        },
                        {
                            "title": "NBC News",
                            "url": "https://www.nbcnews.com/us-news"
                        }
                    ]
                },
                {
                    "name": "Britain",
                    "utils": [
                        {
                            "title": "BBC",
                            "url": "https://bbc.co.uk/"
                        },
                        {
                            "title": "The Guardian",
                            "url": "https://www.theguardian.com/us-news"
                        }
                    ]
                },
                {
                    "name": "Vietnam",
                    "utils": [
                        {
                            "title": "Vnexpress.net",
                            "url": "https://vnexpress.net/"
                        }
                    ]
                },
                {
                    "name": "Business",
                    "utils": [
                        {
                            "title": "Bloomberg",
                            "url": "https://www.bloomberg.com/"
                        }
                    ]
                }
            ]
        }
    ],
}