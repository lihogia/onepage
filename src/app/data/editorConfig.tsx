import type { Category } from '@/app/data/types';

function insertIndexIntoArrayData(arrayList: []) {
    const newArray = arrayList.map((item, index) => {
        return {
            item,
            id: index
        }
    });
}

export function convertToCategoriesInEditor(pCategories: Category[]) {
    const nCategories = pCategories.map((category, index) => {
        const nCategory = {
            id: index,
            ...category
        };
        return nCategory;
    });
    return nCategories;
}