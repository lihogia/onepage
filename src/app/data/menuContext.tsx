
export interface MenuContextItem {
    iconURL: string,
    text: string,
    tooltip: string,
    handle: Function,
    stringIndex: string
}

export const SEPARATOR: MenuContextItem = {
    iconURL: '',
    text: '-',
    tooltip: '',
    handle: () => {},
    stringIndex: ''
}