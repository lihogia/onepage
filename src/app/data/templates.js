export const template001 = {
    "categories": [
        { 
            "name": "Common",
            "subcategories": [
                {
                    "name": "Office Tools",
                    "utils": [
                        {
                            "title": "Google Mail",
                            "url": "https://mail.google.com/mail/u/0/"
                        },
                        {
                            "title": "Google Calendar",
                            "url": "https://calendar.google.com/calendar/b/0/"
                        },
                        {
                            "title": "Google Drive",
                            "url": "https://drive.google.com/u/0/"
                        },
                        {
                            "title": "Microsoft Outlook",
                            "url": "https://outlook.live.com/mail/0/"
                        },
                        {
                            "title": "Outlook Calendar",
                            "url": "https://outlook.live.com/calendar/0/view/month"
                        },
                        {
                            "title": "OneDrive",
                            "url": "https://onedrive.live.com"
                        },
                        {
                            "title": "Yahoo! Mail",
                            "url": "https://mail.yahoo.com/"
                        },
                        {
                            "title": "Proton Mail",
                            "url": "https://mail.proton.me/u/0/inbox"
                        }
                    ] 
                },
                {
                    "name": "Search",
                    "utils": [
                        {
                            "title": "Google",
                            "url": "https://www.google.com/search",
                            "fieldname": "q"
                        },
                        {
                            "title": "Microsoft",
                            "url": "https://www.bing.com/search",
                            "fieldname": "q"
                        },
                        {
                            "title": "Yahoo!",
                            "url": "https://search.yahoo.com/",
                            "fieldname": "p"
                        },
                        {
                            "title": "DuckDuckGo",
                            "url": "https://duckduckgo.com/",
                            "fieldname": "q"
                        },
                        {
                            "title": "百度",
                            "url": "https://www.baidu.com/"
                        }
                    ] 
                },
                {
                    "name": "Dictionaries",
                    "utils": [
                        {
                            "title": "Oxford Dictionary",
                            "url": "https://www.oxfordlearnersdictionaries.com/search/english/direct/",
                            "fieldname": "q"
                        },
                        {
                            "title": 'Cambridge Dictionary',
                            "url": 'https://dictionary.cambridge.org/dictionary/english/e.target.value',
                            "fieldname": 'regexp',
                            "pattern": "e.target.value|ig"
                        },
                        {
                            "title": "Chinese Caligraphy",
                            "url": "https://www.regularcalligraphy.com/character/e.target.value.html",
                            "fieldname": "regexp",
                            "pattern": "e.target.value|ig"
                        }
                    ]
                },
            ]
        },
        {
            "name": "News",
            "subcategories": [
                {
                    "name": "US",
                    "utils": [
                        {
                            "title": "CNN",
                            "url": "https://edition.cnn.com/"
                        },
                        {
                            "title": "Fox News",
                            "url": "https://www.foxnews.com/"
                        },
                        {
                            "title": "Reuters",
                            "url": "https://www.reuters.com/"
                        },
                        {
                            "title": "Nytimes",
                            "url": "https://www.nytimes.com/"
                        },
                        {
                            "title": "CBS News",
                            "url": "https://www.cbsnews.com/"
                        },
                        {
                            "title": "Time",
                            "url": "https://time.com/"
                        },
                        {
                            "title": "npr",
                            "url": "https://www.npr.org/"
                        },
                        {
                            "title": "NBC News",
                            "url": "https://www.nbcnews.com/us-news"
                        },
                    ]
                },
                {
                    "name": "Britain",
                    "utils": [
                        {
                            "title": "BBC",
                            "url": "https://bbc.co.uk/"
                        },
                        {
                            "title": "The Guardian",
                            "url": "https://www.theguardian.com/us-news"
                        },
                    ]
                }
            ]
        },
        {
            "name": "Entertainment",
            "subcategories": [
                {
                    "name": "Videos",
                    "utils": [
                        {
                            "title": "Youtube",
                            "url": "https://www.youtube.com"
                        }
                    ]
                }
            ]
        }
    ]
}