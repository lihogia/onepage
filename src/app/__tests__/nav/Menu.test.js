import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { test001 } from '@/app/data/dataToTest';
import Menu from '../../components/nav/Menu';
import { debug } from 'jest-preview';


describe(Menu, () => {
   
    it('Renders the main menu for Mobile with Categories text', () => {
        const cates = test001.categories;
        render(<Menu categories={cates} selectedIndex={0}/>);
        const theText = screen.getByText(/Categories/i);
        expect(theText).toBeInTheDocument();
    });

    it('Renders the main menu with all Categories are Menu items', () => {
        const cates = test001.categories;
        const { getAllByRole } = render(<Menu categories={cates} selectedIndex={0}/>);
        const aLinks = getAllByRole('link');

        cates.forEach((element, index) => {
            expect(element.name).toEqual(aLinks[index].textContent);
        });
        
    });

    it('Renders the bottom Menu with About, Edit, Config & Buy me a coffee', () => {
        const cates = test001.categories;
        const { getAllByRole } = render(<Menu categories={cates} selectedIndex={0}/>);
        const uLists = getAllByRole('list');

        const uBottoms = uLists.filter(element => {
            return (element.className == "menuBottom");
        });
        expect(uBottoms).toHaveLength(2);
        
        uBottoms.forEach(uElement => {
            const childNodes = uElement.childNodes;
            expect(childNodes[0].querySelector('a').textContent).toEqual('About');
            expect(childNodes[1].querySelector('a').textContent).toEqual('Edit');
            expect(childNodes[2].querySelector('a').textContent).toEqual('Config');
            expect(childNodes[3].querySelector('a').textContent).toEqual('Buy me a coffee');
        });
    });

    it('Renders Menu as snapshot', () => {
        const cates = test001.categories;
        const component = renderer.create(<Menu categories={cates} selectedIndex={1}/>);
        const screen = component.toJSON();
        expect(screen).toMatchSnapshot();
    });

/*
it('Bottom Menu, click on About get the Back to Categories; then click on Back to ... and return to About', async () => {
        //const user = userEvent.setup();

        const cates = test001.categories;
        const component = renderer.create(<Menu categories={cates} selectedIndex={0}/>);
        const screen = component.root;
        const aboutLinks = screen.findAllByProps({children: 'About'})[0];
        //screen.findAllByType('a')[0];
        //console.log(screen.findAllByType('a')[0]);
        console.log(aboutLinks.props.children);

        //expect(component.toJSON()).toMatchSnapshot();

        //const { getAllByRole } = render(<Menu categories={cates} selectedIndex={0}/>);
        

        //console.log(component.toJSON());
        /*

        debug();
        const aboutLinks = .getAllByRole('link', {name: 'About'});
        const firstAboutLink = aboutLinks[0];
        //console.log(firstAboutLink);

        //await user.click(firstAboutLink);

        //console.log(firstAboutLink.textContent);
        
        fireEvent.click(firstAboutLink);
        //fireEvent.mouseDown(firstAboutLink);
        fireEvent(
            firstAboutLink,
            new MouseEvent('click', {
              bubbles: true,
              cancelable: true,
            }),
          )
        //await user.pointer({keys: '[/MouseLeft]', target: firstAboutLink});
        //await user.click(firstAboutLink);
        waitFor(() => {
            console.log(firstAboutLink.textContent);
            expect(firstAboutLink.textContent).toEqual('Back to Categories');
            
        });
        
*/
        /*
        await user.click(firstAboutLink);
        console.log(firstAboutLink.textContent);
        expect(firstAboutLink.textContent).toEqual('About');
        
    });
});
*/
/*
test('Renders the menu correctly, switch mode to About, Edit & Config', () => {
    const component = renderer.create(<Menu categories={test001.categories} selectedIndex={0}/>);

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // About
/*    renderer.act(() => {
        tree.props.onSwitchMode(2);
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();


    // Edit
    renderer.act(() => {
        tree.props.onSwitchMode(1);
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

        // Config
    renderer.act(() => {
        tree.props.onSwitchMode(3);
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // Back to View
    renderer.act(() => {
        tree.props.onSwitchMode(0);
    });
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
*/

});
