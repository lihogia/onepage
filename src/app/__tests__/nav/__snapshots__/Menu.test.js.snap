// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Renders Menu as snapshot 1`] = `
[
  <div
    className="grid1"
  >
    <section
      className="logo"
    >
      <img
        alt="OnePage"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height={90}
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fonepage.png&w=256&q=75"
        srcSet="/_next/image?url=%2Fonepage.png&w=96&q=75 1x, /_next/image?url=%2Fonepage.png&w=256&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={90}
      />
    </section>
    <ul
      className="menu"
    >
      <li
        className="menuItem"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Common
        </a>
      </li>
      <li
        className="menuItemSelected"
      >
        <a
          className=""
          href="#"
        >
          News
        </a>
      </li>
    </ul>
    <ul
      className="menuBottom"
    >
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          About
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Edit
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Config
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="https://www.buymeacoffee.com/lilogia"
          target="_blank"
        >
          Buy me a coffee
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="grid1m"
  >
    <section
      className="logo"
    >
      <img
        alt="OnePage"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height={90}
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fonepage.png&w=256&q=75"
        srcSet="/_next/image?url=%2Fonepage.png&w=96&q=75 1x, /_next/image?url=%2Fonepage.png&w=256&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={90}
      />
    </section>
    <ul
      className="menuTitle"
    >
      <li>
        <a
          href="#"
          onClick={[Function]}
        >
          Categories
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="grid1m_sub_none"
    id="MenuID"
  >
    <ul
      className="menu"
    >
      <li
        className="menuItem"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Common
        </a>
      </li>
      <li
        className="menuItemSelected"
      >
        <a
          className="menuItemSelected"
          href="#"
        >
          News
        </a>
      </li>
    </ul>
    <ul
      className="menuBottom"
    >
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          About
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Edit
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          Config
        </a>
      </li>
      <li
        className="menuItemBottom"
      >
        <a
          href="https://www.buymeacoffee.com/lilogia"
          target="_blank"
        >
          Buy me a coffee
        </a>
      </li>
    </ul>
  </div>,
]
`;
